assets/src/edit-story/components/library/panes/media/virtualList.js

```diff
@@ -0,0 +1,332 @@
+    setScrollInfo((info) => ({ ...info, isScrolling: false }));
+  }, 500);
+
+  const onScrollExec = (evt) => {
+    const { scrollTop, clientHeight, scrollHeight } = evt.target;
+    const from = scrollTop;
+    const to = scrollTop + clientHeight;
+
+    setScrollInfo((info) => ({ ...info, from, to, isScrolling: true }));
+    debouncedIsScrollingFalse();
+
+    const nextPageFetchPadding = 9000;
+    if (isFetching === false && to + nextPageFetchPadding >= scrollHeight) {
+      if (localScopeMedia) {
+        setP((v) => v + 1);
+      } else {
+        loadNextPage();
+      }
+    }
+  };
+
+  useEffect(() => {
+    if (localScopeMedia) {
+      fetchMedia({ pagingNum: p }, gotNewMedia);
+    }
+  }, [gotNewMedia, fetchMedia, p, localScopeMedia]);
+
+  const handleRAFScroll = (evt) => {
+    if (timeoutRef.current) {
+      window.cancelAnimationFrame(timeoutRef.current);
+    }
+    timeoutRef.current = window.requestAnimationFrame(function () {
+      onScrollExec(evt);
+    });
+  };
+  const [debouncedHandleScroll] = useDebouncedCallback((evt) => {
+    onScrollExec(evt);
+  }, 100); //66
+
+  useEffect(() => {
+    const outerNode = outerRef.current;
+    // outerNode.addEventListener('scroll', handleRAFScroll, { passive: true });
+    outerNode.addEventListener('scroll', debouncedHandleScroll, {
+      passive: true,
+    });
+    return () => {
+      // outerNode.removeEventListener('scroll', handleRAFScroll, { passive: true });
+      outerNode.removeEventListener('scroll', debouncedHandleScroll, {
+        passive: true,
+      });
+    };
+  }, [debouncedHandleScroll]);
+
+  useEffect(() => {
+    const outerNode = outerRef.current;
+    const innerNode = innerRef.current;
+    const { scrollTop, clientHeight } = outerNode;
+    setScrollInfo((info) => ({
+      ...info,
+      from: scrollTop,
+      to: scrollTop + clientHeight,
+      height: innerNode.clientHeight,
+    }));
+  }, [resources]);
+
+  const [columns, columnsCumulativeTop] = useMemo(() => {
+    if (resources.length === 0) {
+      return [[], [[], []]];
+    }
+    const columns = [[], []];
+    const columnsCumulativeTop = [[0], [0]];
+
+    const ITEM_MARGIN_BOTTOM = 10;
+
+    for (let i = 0, n = resources.length; i < n; i++) {
+      const resource = resources[i];
+      const { width: originalWidth, height: originalHeight } = resource;
+
+      const oRatio =
+        originalWidth && originalHeight ? originalWidth / originalHeight : 1;
+      const itemHeight = PREVIEW_SIZE / oRatio;
+      const itemTotalHeight = itemHeight + ITEM_MARGIN_BOTTOM;
+
+      const columnsHeights = columnsCumulativeTop.map((c) => c[c.length - 1]);
+      const destColumn = columnsHeights.indexOf(Math.min(...columnsHeights));
+
+      columns[destColumn].push(i);
+      columnsCumulativeTop[destColumn].push(
+        columnsHeights[destColumn] + itemTotalHeight
+      );
+    }
+
+    return [columns, columnsCumulativeTop];
+  }, [resources]);
```

Comment:
I think I understand why you have the distinction between local and global resources and why the reducer contents are spilling to over here. IMHO this indicates that parts of media reducer should move closer to here for better control. However, this also indicates that local or global, we have to recalc all offsets for every time we perform a fetch? Not necessarily a blocker, but I'd like to understand the amount of computations better.